Build after an NQP bump, is broken

## The Problem

Cannot build rakudo after a bump in NQP

## Expected Behavior

Should be able to build rakudo *ALWAYS*

## Actual Behavior

    $ perl Configure.pl --gen-moar --make-install; date
    ===ATTENTION===
      No --prefix supplied, building and installing to /Users/liz/Github/rakudo.moar/install
    Found /Users/liz/Github/rakudo.moar/install/bin/nqp-m (backend moar)
    ===WARNING===
      /Users/liz/Github/rakudo.moar/install/bin/nqp-m version 2020.06-16-g6bdf69921 is outdated, 2020.06-19-g3e928af0b expected.
    remote: Enumerating objects: 8, done.
    remote: Counting objects: 100% (8/8), done.
    remote: Compressing objects: 100% (4/4), done.
    remote: Total 8 (delta 4), reused 5 (delta 4), pack-reused 0
    Unpacking objects: 100% (8/8), done.
    From https://github.com/Raku/nqp
       1cd9594ec..3e928af0b  master     -> origin/master
    Fetching submodule 3rdparty/nqp-configure
    Fetching submodule 3rdparty/nqp-configure
    Previous HEAD position was 6bdf69921 [MoarVM Bump] Brings 20 commits
    HEAD is now at 3e928af0b Merge pull request #646 from patrickbkr/nqp_home_env_precedence
    Building NQP ...
    /usr/local/bin/perl Configure.pl --prefix=/Users/liz/Github/rakudo.moar/install --make-install --git-protocol=https --no-relocatable --no-ignore-errors --silent-build --no-force-rebuild --backends=moar --gen-moar
    Found /Users/liz/Github/rakudo.moar/install/bin/moar version 2020.06-34-gd564695f6, which is new enough.
    Cleaning up ...
    +++ Preparing MOAR build directories
    ++++++ Building MOAR backend
    +++ Generating	gen/moar/stage1/nqpmo.nqp
    +++ Compiling	gen/moar/stage1/nqpmo.moarvm
    +++ Compiling	gen/moar/stage1/ModuleLoader.moarvm
    +++ Generating	gen/moar/stage1/NQPCORE.setting
    +++ Compiling	gen/moar/stage1/NQPCORE.setting.moarvm
    +++ Generating	gen/moar/stage1/QASTNode.nqp
    +++ Compiling	gen/moar/stage1/QASTNode.moarvm
    +++ Generating	gen/moar/stage1/QRegex.nqp
    +++ Compiling	gen/moar/stage1/QRegex.moarvm
    +++ Generating	gen/moar/stage1/NQPHLL.nqp
    +++ Generating	gen/moar/stage1/MASTNodes.nqp
    +++ Compiling	gen/moar/stage1/MASTOps.moarvm
    +++ Compiling	gen/moar/stage1/MASTNodes.moarvm
    +++ Compiling	gen/moar/stage1/NQPHLL.moarvm
    +++ Generating	gen/moar/stage1/QAST.nqp
    +++ Compiling	gen/moar/stage1/QAST.moarvm
    +++ Generating	gen/moar/stage1/NQPP6QRegex.nqp
    +++ Compiling	gen/moar/stage1/NQPP6QRegex.moarvm
    +++ Creating	stage 1 NQP
    +++ Generating	gen/moar/stage2/nqpmo.nqp
    +++ Compiling	gen/moar/stage2/nqpmo.moarvm
    +++ Compiling	gen/moar/stage2/ModuleLoader.moarvm
    +++ Generating	gen/moar/stage2/NQPCORE.setting
    +++ Compiling	gen/moar/stage2/NQPCORE.setting.moarvm
    +++ Generating	gen/moar/stage2/QASTNode.nqp
    +++ Compiling	gen/moar/stage2/QASTNode.moarvm
    +++ Generating	gen/moar/stage2/QRegex.nqp
    +++ Compiling	gen/moar/stage2/QRegex.moarvm
    +++ Generating	gen/moar/stage2/NQPHLL.nqp
    +++ Generating	gen/moar/stage2/MASTNodes.nqp
    +++ Compiling	gen/moar/stage2/MASTOps.moarvm
    +++ Compiling	gen/moar/stage2/MASTNodes.moarvm
    +++ Compiling	gen/moar/stage2/NQPHLL.moarvm
    +++ Generating	gen/moar/stage2/QAST.nqp
    +++ Compiling	gen/moar/stage2/QAST.moarvm
    +++ Generating	gen/moar/stage2/NQPP6QRegex.nqp
    +++ Compiling	gen/moar/stage2/NQPP6QRegex.moarvm
    +++ Creating	stage 2 NQP
    +++ Installing MOAR stage 2 files locally
    +++ Setting up	nqp-m
    +++ Compiling	NQPP5QRegex.moarvm
    +++ Compiling	inst-nqp-m
    +++ Compiling	inst-nqp
    +++ Compiling MOAR default runner
    +++ Preparing MOAR build directories
    ++++++ Building MOAR backend
    +++ Installing MOAR stage 2 files locally
    +++ Compiling	NQPP5QRegex.moarvm
    +++ Creating MOAR installation directories
    +++ Installing into /Users/liz/Github/rakudo.moar/install
    +++ Installing MOAR specifics
    +++ Compiling MOAR default runner
    +++ Installing default runner

    NQP has been built and installed.
    Using /Users/liz/Github/rakudo.moar/install/bin/nqp-m (version 2020.06-19-g3e928af0b / MoarVM 2020.06-34-gd564695f6).
    Cleaning up ...
    +++ Checking for moar NQP version
    +++ Expanding	gen/moar/main-version.nqp
    +++ Generating	gen/moar/rakudo.nqp
    +++ Generating	gen/moar/Grammar.nqp
    +++ Generating	gen/moar/World.nqp
    +++ Generating	gen/moar/ModuleLoader.nqp
    +++ Compiling	blib/Perl6/ModuleLoader.moarvm
    +++ Generating	gen/moar/Ops.nqp
    +++ Compiling	dynext/libperl6_ops_moar.dylib
    +++ Compiling	blib/Perl6/Ops.moarvm
    +++ Generating	gen/moar/Pod.nqp
    +++ Compiling	blib/Perl6/Pod.moarvm
    +++ Compiling	blib/Perl6/World.moarvm
    +++ Generating	gen/moar/Actions.nqp
    +++ Compiling	blib/Perl6/Actions.moarvm
    +++ Compiling	blib/Perl6/Grammar.moarvm
    +++ Generating	gen/moar/Compiler.nqp
    +++ Generating	gen/moar/Optimizer.nqp
    +++ Compiling	blib/Perl6/Optimizer.moarvm
    +++ Compiling	blib/Perl6/Compiler.moarvm
    +++ Compiling	rakudo.moarvm
    +++ Generating	gen/moar/BOOTSTRAP/v6c.nqp
    +++ Generating	gen/moar/Metamodel.nqp
    +++ Compiling	blib/Perl6/Metamodel.moarvm
    +++ Compiling	blib/Perl6/BOOTSTRAP/v6c.moarvm
    +++ Compiling	blib/CORE.c.setting.moarvm
    The following step can take a long time, please be patient.
    Stage start      :   0.000
    Missing or wrong version of dependency 'gen/moar/stage2/QAST.nqp' (from 'gen/moar/BOOTSTRAP/v6c.nqp')
       at <unknown>:1  (/Users/liz/Github/rakudo.moar/install/share/perl6/lib/Perl6/BOOTSTRAP/v6c.moarvm:<dependencies+deserialize>)
     from gen/moar/ModuleLoader.nqp:93  (/Users/liz/Github/rakudo.moar/blib/Perl6/ModuleLoader.moarvm:)
     from gen/moar/ModuleLoader.nqp:76  (/Users/liz/Github/rakudo.moar/blib/Perl6/ModuleLoader.moarvm:load_module)
     from gen/moar/World.nqp:1458  (/Users/liz/Github/rakudo.moar/blib/Perl6/World.moarvm:load_module_early)
     from gen/moar/World.nqp:788  (/Users/liz/Github/rakudo.moar/blib/Perl6/World.moarvm:comp_unit_stage1)
     from gen/moar/Grammar.nqp:840  (/Users/liz/Github/rakudo.moar/blib/Perl6/Grammar.moarvm:comp_unit)
     from gen/moar/Grammar.nqp:555  (/Users/liz/Github/rakudo.moar/blib/Perl6/Grammar.moarvm:TOP)
     from gen/moar/stage2/QRegex.nqp:2258  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/QRegex.moarvm:parse)
     from gen/moar/stage2/NQPHLL.nqp:2300  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:parse)
     from gen/moar/stage2/NQPHLL.nqp:2216  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:execute_stage)
     from gen/moar/stage2/NQPHLL.nqp:2251  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:run)
     from gen/moar/stage2/NQPHLL.nqp:2247  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:)
     from gen/moar/stage2/NQPHLL.nqp:2243  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:compile)
     from gen/moar/stage2/NQPHLL.nqp:1918  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:eval)
     from gen/moar/stage2/NQPHLL.nqp:2153  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:evalfiles)
     from gen/moar/stage2/NQPHLL.nqp:2113  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:command_eval)
     from gen/moar/Compiler.nqp:56  (/Users/liz/Github/rakudo.moar/blib/Perl6/Compiler.moarvm:command_eval)
     from gen/moar/stage2/NQPHLL.nqp:2038  (/Users/liz/Github/rakudo.moar/install/share/nqp/lib/NQPHLL.moarvm:command_line)
     from gen/moar/rakudo.nqp:115  (rakudo.moarvm:MAIN)
     from gen/moar/rakudo.nqp:1  (rakudo.moarvm:<mainline>)
     from <unknown>:1  (rakudo.moarvm:<main>)
     from <unknown>:1  (rakudo.moarvm:<entry>)
    make: *** [blib/CORE.c.setting.moarvm] Error 1
    Command failed (status 512): make
     at /Users/liz/Github/rakudo.moar/3rdparty/nqp-configure/lib/NQP/Config.pm line 34.
        NQP::Config::__ANON__("Command failed (status 512): make\x{a}") called at /Users/liz/Github/rakudo.moar/3rdparty/nqp-configure/lib/NQP/Config.pm line 1483
        NQP::Config::system_or_die("make") called at Configure.pl line 121
    Fri Jul 17 21:58:34 CEST 2020

--- a/src/Perl6/Compiler.nqp
+++ b/src/Perl6/Compiler.nqp
@@ -115,7 +115,8 @@ class Perl6::Compiler is HLL::Compiler {
                 ?? self.config<prefix>
                 !! nqp::substr($execname, 0, nqp::rindex($execname, $sep, nqp::rindex($execname, $sep) - 1));
 
-            $!rakudo-home := nqp::getenvhash()<RAKUDO_HOME>
+            $!rakudo-home := self.cli-options()<rakudo-home>
+                // nqp::getenvhash()<RAKUDO_HOME>
                 // nqp::getenvhash()<PERL6_HOME>
                 // self.config<static-rakudo-home>
                 || $install-dir ~ '/share/perl6';
@@ -182,6 +183,7 @@ and, by default, also executes the compiled code.
   -M module            loads the module prior to running the program
   --target=stage       specify compilation stage to emit
   --optimize=level     use the given level of optimization (0..3)
+  --rakudo-home=path   Override the path of the Rakudo runtime files
   -o, --output=name    specify name of output file
   -v, --version        display version information
   -V                   print configuration summary

--- a/src/main.nqp
+++ b/src/main.nqp
@@ -29,6 +29,7 @@ my @clo := $comp.commandline_options();
 @clo.push('I=s');
 @clo.push('M=s');
 @clo.push('nqp-lib=s');
+@clo.push('rakudo-home=s');
 
 #?if js
 @clo.push('beautify');

--- a/src/vm/moar/runner/main.c
+++ b/src/vm/moar/runner/main.c
@@ -32,7 +32,9 @@ enum {
     FLAG_FULL_CLEANUP,
     FLAG_TRACING,
 
-    OPT_DEBUGPORT
+    OPT_DEBUGPORT,
+
+    OPT_RAKUDO_HOME
 };
 
 static const char *const FLAGS[] = {
@@ -66,6 +68,8 @@ static int parse_flag(const char *arg)
         return (int)(found - FLAGS);
     else if (starts_with(arg, "--debug-port="))
         return OPT_DEBUGPORT;
+    else if (starts_with(arg, "--rakudo-home="))
+        return OPT_RAKUDO_HOME;
     else
         return UNKNOWN_FLAG;
 }
@@ -107,14 +111,20 @@ int retrieve_home(
           char   *exec_dir_path,
           size_t  exec_dir_path_size,
     const char   *check_file,
-    const size_t  check_file_size
+    const size_t  check_file_size,
+          char   *static_home,
+          char   *options_home
 ) {
     char   *check_file_path;
     size_t  home_size;
     int     ret;
     char   *env_home         = getenv(env_var);
 
-    if (env_home) {
+    if (options_home) {
+        *out_home = options_home;
+        home_size = strlen(*out_home);
+    }
+    else if (env_home) {
         home_size = strlen(env_home);
         *out_home = (char*)malloc(home_size + 1);
         strcpy(*out_home, env_home);
@@ -126,6 +136,10 @@ int retrieve_home(
             *(*out_home + home_size - 1) = '\0';
             home_size--;
         }
+    }
+    else if (static_home) {
+        *out_home = static_home;
+        home_size = strlen(*out_home);
     }
     else {
         home_size = exec_dir_path_size + rel_home_size;
@@ -156,28 +170,25 @@ int wmain(int argc, wchar_t *wargv[])
     size_t  exec_path_size;
 
     char   *exec_dir_path_temp;
-#if !(defined(STATIC_NQP_HOME) && defined(STATIC_RAKUDO_HOME)) || defined(_WIN32)
     char   *exec_dir_path;
     size_t  exec_dir_path_size;
-#endif
 
           char   *nqp_home;
           size_t  nqp_home_size;
-#ifndef STATIC_NQP_HOME
+          char   *static_nqp_home     = 0;
     const char    nqp_rel_path[14]    = "/../share/nqp";
     const size_t  nqp_rel_path_size   = 13;
     const char    nqp_check_path[28]  = "/lib/NQPCORE.setting.moarvm";
     const size_t  nqp_check_path_size = 27;
-#endif
 
           char   *rakudo_home;
           size_t  rakudo_home_size;
-#ifndef STATIC_RAKUDO_HOME
+          char   *static_rakudo_home    = 0;
+          char   *option_rakudo_home    = 0;
     const char    perl6_rel_path[16]    = "/../share/perl6";
     const size_t  perl6_rel_path_size   = 15;
     const char    perl6_check_path[22]  = "/runtime/perl6.moarvm";
     const size_t  perl6_check_path_size = 21;
-#endif
 
     char *lib_path[3];
     char *perl6_file;
@@ -250,6 +261,10 @@ int wmain(int argc, wchar_t *wargv[])
                 break;
             }
 
+            case OPT_RAKUDO_HOME:
+                option_rakudo_home = argv[argi] + strlen("--rakudo-home=");
+                break;
+
             default:
             argv[new_argc++] = argv[argi];
         }
@@ -287,7 +302,6 @@ int wmain(int argc, wchar_t *wargv[])
     /* The +1 is the trailing \0 terminating the string. */
     exec_dir_path_temp = (char*)malloc(exec_path_size + 1);
     memcpy(exec_dir_path_temp, exec_path, exec_path_size + 1);
-#if !(defined(STATIC_NQP_HOME) && defined(STATIC_RAKUDO_HOME)) || defined(_WIN32)
 #ifdef _WIN32
     PathRemoveFileSpecA(exec_dir_path_temp);
     exec_dir_path_size = strlen(exec_dir_path_temp);
@@ -297,40 +311,42 @@ int wmain(int argc, wchar_t *wargv[])
     exec_dir_path      = dirname(exec_dir_path_temp);
     exec_dir_path_size = strlen(exec_dir_path);
 #endif
-#endif
 
     /* Retrieve RAKUDO_HOME and NQP_HOME. */
 
 #ifdef STATIC_NQP_HOME
-    nqp_home = STRINGIFY(STATIC_NQP_HOME);
-#else
+    static_nqp_home = STRINGIFY(STATIC_NQP_HOME);
+#endif
     if (!retrieve_home(&nqp_home, nqp_rel_path, nqp_rel_path_size, "NQP_HOME",
-            exec_dir_path, exec_dir_path_size, nqp_check_path, nqp_check_path_size)) {
+            exec_dir_path, exec_dir_path_size, nqp_check_path,
+            nqp_check_path_size, static_nqp_home, 0)) {
         fprintf(stderr, "ERROR: NQP_HOME is invalid: %s\n", nqp_home);
         return EXIT_FAILURE;
     }
-#endif
     nqp_home_size = strlen(nqp_home);
 
 #ifdef STATIC_RAKUDO_HOME
-    rakudo_home = STRINGIFY(STATIC_RAKUDO_HOME);
-#else
+    static_rakudo_home = STRINGIFY(STATIC_RAKUDO_HOME);
+#endif
     /* XXX Isn't it time to move RAKUDO_HOME in front of PERL6_HOME?? */
     if (getenv("PERL6_HOME")) {
-        if (!retrieve_home(&rakudo_home, perl6_rel_path, perl6_rel_path_size, "PERL6_HOME",
-                exec_dir_path, exec_dir_path_size, perl6_check_path, perl6_check_path_size)) {
+        if (!retrieve_home(&rakudo_home, perl6_rel_path, perl6_rel_path_size,
+                "PERL6_HOME", exec_dir_path, exec_dir_path_size,
+                perl6_check_path, perl6_check_path_size, static_rakudo_home,
+                option_rakudo_home)) {
             fprintf(stderr, "ERROR: PERL6_HOME is invalid: %s\n", rakudo_home);
             return EXIT_FAILURE;
         }
     }
     else {
-        if (!retrieve_home(&rakudo_home, perl6_rel_path, perl6_rel_path_size, "RAKUDO_HOME",
-                exec_dir_path, exec_dir_path_size, perl6_check_path, perl6_check_path_size)) {
+        if (!retrieve_home(&rakudo_home, perl6_rel_path, perl6_rel_path_size,
+                "RAKUDO_HOME", exec_dir_path, exec_dir_path_size,
+                perl6_check_path, perl6_check_path_size, static_rakudo_home,
+                option_rakudo_home)) {
             fprintf(stderr, "ERROR: RAKUDO_HOME is invalid: %s\n", rakudo_home);
             return EXIT_FAILURE;
         }
     }
-#endif
     rakudo_home_size = strlen(rakudo_home);
 
     /* Put together the lib paths and perl6_file path. */

--- a/tools/templates/js/Makefile.in
+++ b/tools/templates/js/Makefile.in
@@ -4,7 +4,7 @@
 @bpv(NQP_BASE_FLAGS)@ = --nqp-runtime @bpm(RUNTIME)@ --perl6-runtime @perl6_runtime@ --libpath "@bpm(BLIB)@|||@nqp::libdir@@nfp(/nqp-js-on-js/)@"
 @bpv(NQP_FLAGS)@ = @bpm(NQP_BASE_FLAGS)@ --substagestats --stagestats --source-map
 @bpv(NQP_FLAGS_EXTRA)@ = --execname @bpm(RUNNER)@ --shebang
-@bpv(RUN_RAKUDO)@ = node --max-old-space-size=8192 rakudo.js @bpm(NQP_BASE_FLAGS)@ --source-map
+@bpv(RUN_RAKUDO)@ = node --max-old-space-size=8192 rakudo.js @bpm(NQP_BASE_FLAGS)@ --source-map --rakudo-home=@nfpq($(BASE_DIR)/gen/build_rakudo_home)@
 
 @bpv(CLEANUPS)@ = \
 	@bsm(RAKUDO)@ \

--- a/tools/templates/jvm/Makefile.in
+++ b/tools/templates/jvm/Makefile.in
@@ -5,7 +5,7 @@ JAR     = jar
 NQP_PREFIX  = @nqp_prefix@
 @bpv(NQP)@       = @shquot(@j_nqp@)@
 @bpv(NQP_RR)@    = $(JAVA) -Xss1m -Xms500m -Xmx3000m -cp @q(@nfp(./blib)@@cpsep@@nop($(BLD_NQP_JARS))@@cpsep@rakudo-runtime.jar@cpsep@@nop($(SYSROOT))@@abs2rel(@nqp_classpath@)@)@ nqp
-@bpv(RUN_RAKUDO)@ = $(JAVA) -Xss1m -Xms500m -Xmx3000m -cp @q(@nfp(./blib)@@cpsep@@nop($(BLD_NQP_JARS))@@cpsep@rakudo-runtime.jar@cpsep@rakudo.jar@cpsep@@nop($(SYSROOT))@@abs2rel(@nqp_classpath@)@)@ perl6
+@bpv(RUN_RAKUDO)@ = $(JAVA) -Xss1m -Xms500m -Xmx3000m -cp @q(@nfp(./blib)@@cpsep@@nop($(BLD_NQP_JARS))@@cpsep@rakudo-runtime.jar@cpsep@rakudo.jar@cpsep@@nop($(SYSROOT))@@abs2rel(@nqp_classpath@)@)@ perl6 --rakudo-home=@nfpq($(BASE_DIR)/gen/build_rakudo_home)@
 
 @bpv(DEBUG_RUNNER)@ = rakudo-debug-@backend_abbr@@bpm(RUNNER_SUFFIX)@
 

--- a/tools/templates/moar/Makefile.in
+++ b/tools/templates/moar/Makefile.in
@@ -11,7 +11,7 @@ MOAR_PREFIX = @nfp(@moar::prefix@)@
 MOAR        = @nfpq(@moar::bindir@/moar@moar::exe@)@
 @bpv(NQP)@       = @nfpq(@m_nqp@)@
 @bpv(NQP_RR)@    = @bpm(NQP)@
-@bpv(RUN_RAKUDO)@ = $(MOAR) --libpath=@nfpq($(BASE_DIR)/blib)@ --libpath=@q(@bpm(NQP_LIBDIR)@)@ @bsm(RAKUDO)@ --nqp-lib=@nfpq($(BASE_DIR)/blib)@
+@bpv(RUN_RAKUDO)@ = $(MOAR) --libpath=@nfpq($(BASE_DIR)/blib)@ --libpath=@q(@bpm(NQP_LIBDIR)@)@ @bsm(RAKUDO)@ --nqp-lib=@nfpq($(BASE_DIR)/blib)@ --rakudo-home=@nfpq($(BASE_DIR)/gen/build_rakudo_home)@
 
 @bpv(RUNNER_SUFFIX)@ = @moar::exe@
 

